apps/nextjs/src/app/api/chat/chatHandler.ts

```diff
@@ -167,59 +171,227 @@ async function generateChatStream(
+  output: unknown,
+  chatId: string,
+): { messages: Message[]; lessonPlan: LooseLessonPlan } {
+  let messages: Message[] = [];
+  let lessonPlan: LooseLessonPlan = {};
+
+  try {
+    const parsedOutput =
+      typeof output === "string" ? JSON.parse(output) : output;
+
+    if (hasMessages(parsedOutput) && isValidMessages(parsedOutput.messages)) {
+      messages = parsedOutput.messages as Message[];
+    }
+
+    if (
+      hasLessonPlan(parsedOutput) &&
+      isValidLessonPlan(parsedOutput.lessonPlan)
+    ) {
+      lessonPlan = parsedOutput.lessonPlan as LooseLessonPlan;
+    }
+  } catch (error) {
+    log.error(`Error parsing output for chat ${chatId}`, error);
+  }
+
+  return { messages, lessonPlan };
+}
+
+async function loadChatDataFromDatabase(
+  chatId: string,
+  userId: string,
+): Promise<{ messages: Message[]; lessonPlan: LooseLessonPlan }> {
+  try {
+    const chat = await prisma.appSession.findUnique({
+      where: { id: chatId },
+      select: {
+        id: true,
+        userId: true,
+        output: true,
+      },
+    });
+
+    if (!chat) {
+      log.info(`No existing chat found for id: ${chatId}`);
+      return { messages: [], lessonPlan: {} };
+    }
+
+    verifyChatOwnership(chat, userId, chatId);
+
+    const { messages, lessonPlan } = parseChatOutput(chat.output, chatId);
+
+    log.info(
+      `Loaded ${messages.length} messages and lesson plan for chat ${chatId}`,
+    );
+    return { messages, lessonPlan };
+  } catch (error) {
+    log.error(`Error loading chat data for chat ${chatId}`, error);
+    return { messages: [], lessonPlan: {} };
+  }
+}
+
+function extractLatestUserMessage(frontendMessages: Message[]): Message | null {
+  if (!frontendMessages || frontendMessages.length === 0) {
+    return null;
+  }
+
+  for (let i = frontendMessages.length - 1; i >= 0; i--) {
+    const message = frontendMessages[i];
+    if (message && message.role === "user") {
+      return message;
+    }
+  }
+
+  return null;
+}
+
+function prepareMessages(
+  dbMessages: Message[],
+  frontendMessages: Message[],
+  chatId: string,
+): Message[] {
+  const latestUserMessage = extractLatestUserMessage(frontendMessages);
+
+  let messages = [...dbMessages];
+  if (
+    latestUserMessage &&
+    !messages.some((m) => m.id === latestUserMessage.id)
+  ) {
+    messages.push(latestUserMessage);
+    log.info(`Appended new user message to history for chat ${chatId}`);
+  }
+
+  return messages;
+}
+
+async function createAilaInstance({
+  config,
+  options,
+  chatId,
+  userId,
+  messages,
+  lessonPlan,
+  llmService,
+  moderationAiClient,
+  threatDetectors,
+}: {
+  config: Config;
+  options: AilaOptions;
+  chatId: string;
+  userId: string | undefined;
+  messages: Message[];
+  lessonPlan: LooseLessonPlan;
+  llmService: ReturnType<typeof getFixtureLLMService>;
+  moderationAiClient: ReturnType<typeof getFixtureModerationOpenAiClient>;
+  threatDetectors: AilaThreatDetector[];
```

Comment:
I would argue that this type is too large to be inline, though that's a style preference so a nitpick rather than a blocker!

Reply:
Agreed - made that a separate type. TBH this file is getting pretty big and should probably be broken up